### PR TITLE
Binary attributes

### PR DESCRIPTION
**Pull request by romainthibaux**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds end-to-end support for binary per-box attributes in YOLO11/pose workflows, along with custom training/export utilities and minor logging tweaks.

### 📊 Key Changes
- ➕ Introduces `training_utils` helpers (`training_utils.py`, `simple_training.py`, `export.py`) to prepare Verdant-style datasets (classes + optional attributes), run YOLO-based training, and export ONNX checkpoints at multiple image sizes.
- 🧩 Extends dataset config and parsing to handle `attributes`/`na` in data YAML, propagating attribute tensors alongside `cls`, `bboxes`, and `keypoints` (including `ignore_kpt` flags for flexible keypoint supervision).
- 🧠 Updates YOLO11 detect and pose heads to support an additional `na` attribute dimension, adjusting channel counts, parsing in `parse_model`, and ensuring postprocessing works with combined class+attribute outputs.
- 📦 Enhances loss functions: detection loss now consumes attribute targets and reports an `attr_loss`, and pose loss is refactored to compute attribute loss plus conditional keypoint loss using the new `ignore_kpt` signal.
- 🧪 Integrates attributes into training/validation: validators now track `attr_loss`, datasets carry `attributes` through label caching and filtering, and the core trainer asserts the presence of `ignore_kpt` in batches.
- ⚙️ Adds an `attr` hyperparameter to `default.yaml` to weight attribute loss, and augments `yolo11-pose.yaml` with an `na` field used by the Pose head.
- 📈 Adjusts W&B callbacks and settings to default-enable wandb logging with a simplified `wb.init` call.

### 🎯 Purpose & Impact
- 🎯 Enables binary attribute prediction per detection (conditioned on class), which is useful for tasks that require richer labels than pure class IDs, such as presence/absence of properties or states.
- 🔗 Keeps attributes tightly integrated into the existing training pipeline (data loading, TAL assignment, losses, validators), reducing friction for users who want to extend YOLO11/pose to multi-task setups.
- 🧪 Improves robustness of pose training with flexible keypoint supervision via `ignore_kpt`, allowing images without keypoint annotations to coexist with fully annotated ones.
- 🚀 Provides ready-to-use scripts and configs for Verdant-like datasets, making it easier to bootstrap custom experiments and export trained models to ONNX for deployment.
- 📊 Exposes attribute loss as a first-class metric, giving users better visibility into multi-task performance during training and validation.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @romainthibaux, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review helper, and a Ultralytics engineer will be with you soon. In the meantime, to ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the **Update branch** button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing `I have read the CLA Document and I sign the CLA` in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: training_utils/export.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from training_utils import (
+    GiveModel
+)
+import os
+import argparse
+
+
+def Export(checkpoint_file_path):
```
Comment (UltralyticsAssistant) [inline training_utils/export.py:8]: 💡 **MEDIUM**: `Export` mixes library-style behavior with CLI behavior: it calls `exit(1)` and prints directly. This makes it hard to reuse programmatically and complicates error handling. Prefer raising an exception in `Export` and only calling `sys.exit` in the `__main__` block.

File: training_utils/export.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    if os.path.exists(checkpoint_file_path):
+        model = GiveModel(checkpoint_file_path)
```
Comment (UltralyticsAssistant) [inline training_utils/export.py:10]: 💡 **MEDIUM**: `GiveModel(checkpoint_file_path)` can return `None` if the checkpoint exists but `YOLO` fails to load. In that case, `model.export(...)` will raise an AttributeError. Consider checking the return value and failing fast with a clear message.

File: training_utils/export.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    else:
+        print(f"[ERROR] : Model {checkpoint_file_path} does not exists")
+        exit(1)
+
+    prefix = os.path.splitext(os.path.basename(checkpoint_file_path))[0]
+
+    base_path, _ = os.path.split(checkpoint_file_path)
+    path = model.export(format="onnx", imgsz=[2144, 768], opset=12)
```
Comment (UltralyticsAssistant) [inline training_utils/export.py:18]: 💡 **MEDIUM**: Using `os.system(f"mv {path} ...")` is brittle and unsafe when paths contain spaces or shell metacharacters, and it also ignores errors. Prefer using `os.replace`/`shutil.move` and checking for exceptions instead of shelling out.

File: training_utils/export.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    os.system(f"mv {path} {base_path}/{prefix}_full_height.onnx")
+
+    path = model.export(format="onnx", imgsz=[2144, 4096], opset=12)
+    os.system(f"mv {path} {base_path}/{prefix}_full_frame.onnx")
+
+    path = model.export(format="onnx", imgsz=[768, 768], opset=12)
```
Comment (UltralyticsAssistant) [inline training_utils/export.py:24]: 📝 **LOW**: The third export (`imgsz=[768, 768]`) computes `path` but never renames or returns it. This looks like a bug or an unfinished feature; either move/return the file or remove the export to avoid confusion.

File: training_utils/simple_training.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+)
+from training_utils import (
+    dataset_yaml_path,
+    coco_classes_file,
+    training_task,
+    experiment_name,
+)
+import os
+import argparse
+from export import Export
```
Comment (UltralyticsAssistant) [inline training_utils/simple_training.py:15]: ⚠️ **HIGH**: `from export import Export` performs an absolute import, which will resolve differently depending on the working directory and `PYTHONPATH`. Inside this package, this will likely not import `training_utils/export.py` as intended and can break when installed as a library. Use a package-qualified or explicit relative import instead.

**Suggested change:**
```suggestion
from training_utils.export import Export
```

File: training_utils/simple_training.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            exit(1)
+    else:
+        model = YOLO(GetModelYaml(training_task))  # Initialize model
+
+    if args.disable_wandb:
+        os.environ['WANDB_MODE'] = 'disabled'
+
+    model.train(
+        task=training_task,
+        data="verdant.yaml",
```
Comment (UltralyticsAssistant) [inline training_utils/simple_training.py:48]: 💡 **MEDIUM**: `data="verdant.yaml"` assumes the working directory contains this file, but `PrepareDataset` writes `dataset_yaml_path`, which may point elsewhere (e.g. an absolute path). This mismatch can cause the trainer to pick up the wrong YAML. Consider passing `dataset_yaml_path` directly to `model.train`.

**Suggested change:**
```suggestion
        data=dataset_yaml_path,
```

File: training_utils/simple_training.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        epochs=args.epochs,
+        flipud=0.5,
+        fliplr=0.5,
+        scale=0.2,
+        mosaic=0.0, # Please set this to 0.0 TODO: Fix the issue with mosaic and keypoint detection
+        imgsz=768,
+        seed=1,
+        batch=args.batch_size,
+        name=experiment_name,
+        device=[0, 1, 2, 3, 4, 5, 6, 7],
```
Comment (UltralyticsAssistant) [inline training_utils/simple_training.py:61]: 💡 **MEDIUM**: Hard-coding `device=[0, 1, 2, 3, 4, 5, 6, 7]` assumes 8 visible GPUs and will crash or behave unexpectedly on CPU-only or smaller GPU setups. Prefer `device=0`, `device="cuda"`, or configuring devices via arguments or environment variables.

File: training_utils/training_utils.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def PrepareDataset(coco_classes_file, dataset_yaml, training_task):
+    classes = []
+    attributes = []
+    with open(coco_classes_file, "r") as f:
+        for line in f.readlines():
+            classes.append(line.strip("\n").strip(" "))
+
+    if os.path.exists(coco_attributes_file):
```
Comment (UltralyticsAssistant) [inline training_utils/training_utils.py:21]: 📝 **LOW**: `PrepareDataset` ignores its `coco_classes_file` parameter for attributes and instead uses the module-level `coco_attributes_file`. This is slightly confusing and makes the function less self-contained. Either accept an attributes file path as an argument or consistently use globals.

File: training_utils/training_utils.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        f.write("test: images/test\n\n")
+        f.write("names:\n")
+        for i in range(len(classes)):
+            f.write(f"  {i}: {classes[i]}\n")
+        if len(attributes) > 0:
+            f.write("\nattributes:\n")
+            for i in range(len(attributes)):
+                f.write(f"  {i}: {attributes[i]}\n")
+        if training_task == "pose":
+            f.write("\nkpt_shape: [1, 3]\n")  # enforce keypoint shape to [1, 3] for pose models
```
Comment (UltralyticsAssistant) [inline training_utils/training_utils.py:39]: 📝 **LOW**: The comment in the pose branch hard-codes `kpt_shape: [1, 3]` and `flip_idx: [0]`. If these values ever change, they will diverge from the actual YAML. Consider driving these from configuration or constants to avoid hidden coupling.

File: training_utils/training_utils.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        return YOLO(ckpt_path)
+    print(f"[ERROR] : Model {ckpt_path} does not exists")
+    return None
+
+
+def GetLatestWeightsDir():
+    base_dir = f"{runs_directory}/{training_task}"
+
+    # Get all entries in the directory given by path
+    entries = os.listdir(base_dir)
```
Comment (UltralyticsAssistant) [inline training_utils/training_utils.py:64]: ⚠️ **HIGH**: `GetLatestWeightsDir` calls `os.listdir(base_dir)` without checking whether `base_dir` exists. When the runs directory or task subdirectory is missing (e.g. first-time run, misconfigured `RUNS_DIR`), this will raise an uncaught `FileNotFoundError`. Add an existence check and a clear error message or fallback.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Major concerns:
- Backward compatibility: `na` handling in `parse_model` and the heads is currently unsafe for existing configs and checkpoints that do not define `na`; defaulting `na` to 0 is essential.
- Dataset and batch structure: new `attributes` and `ignore_kpt` fields must be present and well-typed across the entire pipeline; otherwise `update_labels`, the trainer assertion, and pose loss can fail or behave unexpectedly.
- Utility scripts: imports and error handling in `training_utils` (`Export`, `simple_training`, `GetLatestWeightsDir`) should be hardened to avoid brittle behavior in real training runs.
- Logging and configuration: enabling WANDB by default and changing W&B project/name behavior may surprise users; consider making this opt-in or restoring previous sanitization.

Once the `na` defaults, dataset keys, and a few robustness issues are addressed, the overall design for binary attributes and keypoint ignoring looks solid and well-integrated with YOLO11/pose workflows.

💬 Posted 10 inline comments
